File: app/services/auth.service.ts
import { firebase } from '@nativescript/firebase-core';
import '@nativescript/firebase-auth';
import { User } from '../models/user.model';

export class AuthService {
    private static instance: AuthService;
    
    static getInstance(): AuthService {
        if (!AuthService.instance) {
            AuthService.instance = new AuthService();
        }
        return AuthService.instance;
    }

    async login(username: string, password: string): Promise<User | null> {
        try {
            const auth = firebase().auth();
            const userCredential = await auth.signInWithEmailAndPassword(username, password);
            
            // Get user data from Realtime Database
            const database = firebase().database();
            const userData = await database.ref(`users/${userCredential.user.uid}`).once('value');
            
            return userData.val() as User;
        } catch (error) {
            console.error('Login error:', error);
            return null;
        }
    }

    async createUser(user: Omit<User, 'id'>): Promise<boolean> {
        try {
            const auth = firebase().auth();
            const userCredential = await auth.createUserWithEmailAndPassword(user.username, user.password);
            
            const database = firebase().database();
            await database.ref(`users/${userCredential.user.uid}`).set({
                ...user,
                id: userCredential.user.uid
            });
            
            return true;
        } catch (error) {
            console.error('Create user error:', error);
            return false;
        }
    }

    async logout(): Promise<void> {
        const auth = firebase().auth();
        await auth.signOut();
    }
}

File: app/src/main/kotlin/org/censusapp/services/AuthService.kt
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import kotlinx.coroutines.tasks.await

class AuthService private constructor() {
    private val auth = FirebaseAuth.getInstance()
    private val database = FirebaseDatabase.getInstance()

    suspend fun login(username: String, password: String): User? {
        return try {
            val result = auth.signInWithEmailAndPassword(username, password).await()
            val snapshot = database.reference
                .child("users")
                .child(result.user?.uid ?: return null)
                .get()
                .await()
            
            snapshot.getValue(User::class.java)
        } catch (e: Exception) {
            null
        }
    }

    suspend fun createUser(user: User): Boolean {
        return try {
            val result = auth.createUserWithEmailAndPassword(user.username, user.password).await()
            database.reference
                .child("users")
                .child(result.user?.uid ?: return false)
                .setValue(user.copy(id = result.user?.uid ?: ""))
                .await()
            true
        } catch (e: Exception) {
            false
        }
    }

    fun logout() {
        auth.signOut()
    }

    companion object {
        @Volatile
        private var instance: AuthService? = null

        fun getInstance() = instance ?: synchronized(this) {
            instance ?: AuthService().also { instance = it }
        }
    }
}